src/nni_engine.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright 2019-2022 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+//
+
+#include "nni_engine.hpp"
+
+#include "gp_engine.hpp"
+#include "gp_dag.hpp"
+
+#include "bitset.hpp"
+#include "subsplit_dag.hpp"
+#include "nni_operation.hpp"
+#include "graft_dag.hpp"
+#include "sugar.hpp"
+
+NNIEngine::NNIEngine(GPDAG &dag, GPEngine &gp_engine)
+    : dag_(dag),
+      graft_dag_(std::make_unique<GraftDAG>(dag)),
+      gp_engine_(gp_engine),
+      run_count_(0) {}
+
+// ** Runners
+
+void NNIEngine::Run() {
+  RunInit();
+  // Loop until no more eligible NNIs are found.
+  while (GetAdjacentNNICount() > 0) {
+    RunMainLoop();
+    RunPostLoop();
+  }
+}
+
+void NNIEngine::RunInit() {
+  // Initialize Adjacent NNIs based on starting state of DAG.
+  ClearAdjacentNNIs();
+  SyncAdjacentNNIsWithDAG();
+  // Initialize GPEngine for current DAG size
+  ResizeGPEngineForDAG();
+  PrepGPEngineForLikelihoods();
+}
+
+void NNIEngine::RunMainLoop() {
+  // (1) Add all adjacent NNIs to the graftDAG.
+  GraftAdjacentNNIsToDAG();
+  // (2) Compute marginal likelihoods for each adjacent NNI.
+  EvaluateAdjacentNNIs();
+  // (3) Select whether to accept or reject Adjacent NNIs via filter.
+  FilterAdjacentNNIsToAcceptedNNIs();
+  // (4) Add Accepted NNIs to Permanent DAG.
+  AddAcceptedNNIsToDAG();
```

Comment:
I realize this is sort of peudocode, but don't we want to remove grafts for every added NNI?

Reply:
Currently, I just have it so at the end of each pass, all the nodes and edges are cleared from the graft.  It bypasses doing any sort of manipulation on the individual nodes and edges in the graft.